File: Devken.CBC.SchoolManagement.Application/DTOs/Payments/payments.cs
using Devken.CBC.SchoolManagement.Domain.Enums;
using System.ComponentModel.DataAnnotations;

namespace Devken.CBC.SchoolManagement.Application.DTOs.Payments
{
    // ══════════════════════════════════════════════════════════════════════════
    //  RESPONSE
    // ══════════════════════════════════════════════════════════════════════════

    public class PaymentResponseDto
    {
        public Guid Id { get; set; }
        public Guid TenantId { get; set; }

        public string PaymentReference { get; set; } = null!;
        public string? ReceiptNumber { get; set; }

        public Guid StudentId { get; set; }
        public string? StudentName { get; set; }
        public string? AdmissionNumber { get; set; }

        public Guid InvoiceId { get; set; }
        public string? InvoiceNumber { get; set; }

        public Guid? ReceivedBy { get; set; }
        public string? ReceivedByName { get; set; }

        public DateTime PaymentDate { get; set; }
        public DateTime? ReceivedDate { get; set; }
        public decimal Amount { get; set; }
        public string PaymentMethod { get; set; } = null!;
        public string StatusPayment { get; set; } = null!;
        public string? TransactionReference { get; set; }
        public string? Description { get; set; }
        public string? Notes { get; set; }

        public string? MpesaCode { get; set; }
        public string? PhoneNumber { get; set; }

        public string? BankName { get; set; }
        public string? AccountNumber { get; set; }
        public string? ChequeNumber { get; set; }
        public DateTime? ChequeClearanceDate { get; set; }

        public Guid? ReversedFromPaymentId { get; set; }
        public bool IsReversal { get; set; }
        public string? ReversalReason { get; set; }

        public bool IsCompleted { get; set; }
        public bool IsMpesa { get; set; }

        // BaseEntity<TId> audit fields
        public DateTime CreatedOn { get; set; }
        public DateTime UpdatedOn { get; set; }
        public Guid? CreatedBy { get; set; }
        public Guid? UpdatedBy { get; set; }
    }

    // ══════════════════════════════════════════════════════════════════════════
    //  CREATE
    // ══════════════════════════════════════════════════════════════════════════

    public class CreatePaymentDto
    {
        public Guid? TenantId { get; set; }   // resolved server-side for non-SuperAdmin

        [Required] public Guid StudentId { get; set; }
        [Required] public Guid InvoiceId { get; set; }
        public Guid? ReceivedBy { get; set; }

        [Required] public DateTime PaymentDate { get; set; }
        public DateTime? ReceivedDate { get; set; }

        [Required]
        [Range(0.01, double.MaxValue, ErrorMessage = "Amount must be greater than zero.")]
        public decimal Amount { get; set; }

        [Required] public PaymentMethod PaymentMethod { get; set; } = PaymentMethod.Cash;
        public PaymentStatus StatusPayment { get; set; } = PaymentStatus.Completed;

        [MaxLength(100)] public string? TransactionReference { get; set; }
        [MaxLength(500)] public string? Description { get; set; }
        [MaxLength(1000)] public string? Notes { get; set; }

        [MaxLength(20)] public string? MpesaCode { get; set; }
        [MaxLength(20)] public string? PhoneNumber { get; set; }

        [MaxLength(100)] public string? BankName { get; set; }
        [MaxLength(50)] public string? AccountNumber { get; set; }
        [MaxLength(50)] public string? ChequeNumber { get; set; }
        public DateTime? ChequeClearanceDate { get; set; }
    }

    // ══════════════════════════════════════════════════════════════════════════
    //  UPDATE  (partial — null = leave unchanged)
    // ══════════════════════════════════════════════════════════════════════════

    public class UpdatePaymentDto
    {
        public DateTime? PaymentDate { get; set; }
        public DateTime? ReceivedDate { get; set; }

        [Range(0.01, double.MaxValue, ErrorMessage = "Amount must be greater than zero.")]
        public decimal? Amount { get; set; }

        public PaymentMethod? PaymentMethod { get; set; }
        public PaymentStatus? StatusPayment { get; set; }
        public Guid? ReceivedBy { get; set; }

        [MaxLength(100)] public string? TransactionReference { get; set; }
        [MaxLength(500)] public string? Description { get; set; }
        [MaxLength(1000)] public string? Notes { get; set; }

        [MaxLength(20)] public string? MpesaCode { get; set; }
        [MaxLength(20)] public string? PhoneNumber { get; set; }

        [MaxLength(100)] public string? BankName { get; set; }
        [MaxLength(50)] public string? AccountNumber { get; set; }
        [MaxLength(50)] public string? ChequeNumber { get; set; }
        public DateTime? ChequeClearanceDate { get; set; }
    }

    // ══════════════════════════════════════════════════════════════════════════
    //  REVERSAL
    // ══════════════════════════════════════════════════════════════════════════

    public class ReversePaymentDto
    {
        [Required]
        [MaxLength(500)]
        public string ReversalReason { get; set; } = null!;

        public Guid? ReceivedBy { get; set; }
    }

    // ══════════════════════════════════════════════════════════════════════════
    //  BULK  (CBC — whole class / grade payment run)
    // ══════════════════════════════════════════════════════════════════════════

    public class BulkPaymentItemDto
    {
        [Required] public Guid StudentId { get; set; }
        [Required] public Guid InvoiceId { get; set; }

        [Required]
        [Range(0.01, double.MaxValue, ErrorMessage = "Amount must be greater than zero.")]
        public decimal Amount { get; set; }

        [MaxLength(20)] public string? MpesaCode { get; set; }
        [MaxLength(20)] public string? PhoneNumber { get; set; }
        [MaxLength(100)] public string? TransactionReference { get; set; }
        [MaxLength(500)] public string? Notes { get; set; }
    }

    public class BulkPaymentDto
    {
        public Guid? TenantId { get; set; }   // resolved server-side for non-SuperAdmin

        [Required] public DateTime PaymentDate { get; set; }
        [Required] public PaymentMethod PaymentMethod { get; set; } = PaymentMethod.Cash;
        public PaymentStatus StatusPayment { get; set; } = PaymentStatus.Completed;

        public Guid? ReceivedBy { get; set; }

        [MaxLength(500)] public string? Description { get; set; }
        [MaxLength(100)] public string? BankName { get; set; }
        [MaxLength(50)] public string? AccountNumber { get; set; }

        [Required]
        [MinLength(1, ErrorMessage = "At least one payment item is required.")]
        public List<BulkPaymentItemDto> Payments { get; set; } = new();
    }

    // ══════════════════════════════════════════════════════════════════════════
    //  BULK RESULT
    // ══════════════════════════════════════════════════════════════════════════

    public class BulkPaymentResultDto
    {
        public int TotalRequested { get; set; }
        public int Succeeded { get; set; }
        public int Failed { get; set; }
        public decimal TotalAmountPosted { get; set; }

        public List<PaymentResponseDto> CreatedPayments { get; set; } = new();
        public List<BulkPaymentErrorDto> Errors { get; set; } = new();
    }

    public class BulkPaymentErrorDto
    {
        public Guid StudentId { get; set; }
        public Guid InvoiceId { get; set; }
        public string Reason { get; set; } = null!;
    }
}

File: Devken.CBC.SchoolManagement.API/Devken.CBC.SchoolManagment.UI/School-System-UI/src/app/payments/types/payments.ts
// ═══════════════════════════════════════════════════════════════════
// payments.ts  — Types, enums and helpers
// Mirrors the C# DTOs in Payments DTOs file
// ═══════════════════════════════════════════════════════════════════

// ── Enums (string-based, matching C# enum names) ─────────────────

export type PaymentMethod =
  | 'Cash'
  | 'Mpesa'
  | 'BankTransfer'
  | 'Cheque'
  | 'Card'
  | 'Online';

export type PaymentStatus =
  | 'Pending'
  | 'Completed'
  | 'Failed'
  | 'Refunded'
  | 'Cancelled'
  | 'Reversed';

// ── Enum → integer maps (sent to API) ────────────────────────────

export const PaymentMethodValue: Record<PaymentMethod, number> = {
  Cash:         0,
  Mpesa:        1,
  BankTransfer: 2,
  Cheque:       3,
  Card:         4,  // CreditCard = 4
  Online:       5,  // OnlinePortal = 5
};

export const PaymentStatusValue: Record<PaymentStatus, number> = {
  Pending:   0,
  Completed: 1,
  Failed:    2,
  Refunded:  3,
  Cancelled: 4,
  Reversed:  5,
};
export interface PaymentPagedResultDto {
    schoolCount: string;
    items:              PaymentResponseDto[];
    totalCount:         number;
    page:               number;
    pageSize:           number;
    totalPages:         number;
    totalCollected:     number;
    totalReversed:      number;
    netAvailable:       number;
    pendingCount:       number;
    mpesaCount:         number;
    totalReversalCount: number;
}
// ── Response DTO ──────────────────────────────────────────────────

export interface PaymentResponseDto {
  id:                    string;
  tenantId:              string;

  paymentReference:      string;
  receiptNumber?:        string;

  studentId:             string;
  studentName?:          string;
  admissionNumber?:      string;

  invoiceId:             string;
  invoiceNumber?:        string;

  receivedBy?:           string;
  receivedByName?:       string;

  paymentDate:           string;
  receivedDate?:         string;
  amount:                number;
  paymentMethod:         PaymentMethod;
  statusPayment:         PaymentStatus;
  transactionReference?: string;
  description?:          string;
  notes?:                string;

  mpesaCode?:            string;
  phoneNumber?:          string;

  bankName?:             string;
  accountNumber?:        string;
  chequeNumber?:         string;
  chequeClearanceDate?:  string;

  reversedFromPaymentId?: string;
  isReversal:            boolean;
  reversalReason?:       string;

  isCompleted:           boolean;
  isMpesa:               boolean;

  createdOn:             string;
  updatedOn:             string;
  createdBy?:            string;
  updatedBy?:            string;
}

// ── Create DTO ────────────────────────────────────────────────────

export interface CreatePaymentDto {
  tenantId?:             string;

  studentId:             string;
  invoiceId:             string;
  receivedBy?:           string;

  paymentDate:           string;
  receivedDate?:         string;

  amount:                number;
  paymentMethod:         number;  // integer sent to API
  statusPayment?:        number;  // integer sent to API

  transactionReference?: string;
  description?:          string;
  notes?:                string;

  mpesaCode?:            string;
  phoneNumber?:          string;

  bankName?:             string;
  accountNumber?:        string;
  chequeNumber?:         string;
  chequeClearanceDate?:  string;
}

// ── Update DTO (all fields optional — partial update) ─────────────

export interface UpdatePaymentDto {
  paymentDate?:          string;
  receivedDate?:         string;

  amount?:               number;
  paymentMethod?:        number;  // integer sent to API
  statusPayment?:        number;  // integer sent to API
  receivedBy?:           string;

  transactionReference?: string;
  description?:          string;
  notes?:                string;

  mpesaCode?:            string;
  phoneNumber?:          string;

  bankName?:             string;
  accountNumber?:        string;
  chequeNumber?:         string;
  chequeClearanceDate?:  string;
}

// ── Reversal DTO ──────────────────────────────────────────────────

export interface ReversePaymentDto {
  reversalReason: string;
  receivedBy?:    string;
}

// ── Bulk DTO ──────────────────────────────────────────────────────

export interface BulkPaymentItemDto {
  studentId:             string;
  invoiceId:             string;
  amount:                number;
  mpesaCode?:            string;
  phoneNumber?:          string;
  transactionReference?: string;
  notes?:                string;
}

export interface BulkPaymentDto {
  tenantId?:      string;
  paymentDate:    string;
  paymentMethod:  number;   // integer sent to API
  statusPayment?: number;   // integer sent to API
  receivedBy?:    string;
  description?:   string;
  bankName?:      string;
  accountNumber?: string;
  payments:       BulkPaymentItemDto[];
}

// ── Bulk Result ───────────────────────────────────────────────────

export interface BulkPaymentResultDto {
  totalRequested:    number;
  succeeded:         number;
  failed:            number;
  totalAmountPosted: number;
  createdPayments:   PaymentResponseDto[];
  errors:            BulkPaymentErrorDto[];
}

export interface BulkPaymentErrorDto {
  studentId: string;
  invoiceId: string;
  reason:    string;
}

// ── Summary DTO ───────────────────────────────────────────────────

export interface PaymentSummaryDto {
  totalPayments:   number;
  totalAmount:     number;
  completedCount:  number;
  completedAmount: number;
  pendingCount:    number;
  pendingAmount:   number;
  reversalCount:   number;
  reversalAmount:  number;
  byMethod:        { method: PaymentMethod; count: number; amount: number }[];
}

// ── Label / colour helpers ────────────────────────────────────────

export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  Cash:         'Cash',
  Mpesa:        'M-Pesa',
  BankTransfer: 'Bank Transfer',
  Cheque:       'Cheque',
  Card:         'Card',
  Online:       'Online',
};

export const PAYMENT_METHOD_ICONS: Record<PaymentMethod, string> = {
  Cash:         'payments',
  Mpesa:        'phone_android',
  BankTransfer: 'account_balance',
  Cheque:       'description',
  Card:         'credit_card',
  Online:       'language',
};

export const PAYMENT_METHOD_COLORS: Record<PaymentMethod, string> = {
  Cash:         'emerald',
  Mpesa:        'green',
  BankTransfer: 'blue',
  Cheque:       'amber',
  Card:         'violet',
  Online:       'indigo',
};

export const PAYMENT_STATUS_COLORS: Record<PaymentStatus, string> = {
  Pending:   'amber',
  Completed: 'green',
  Failed:    'red',
  Refunded:  'blue',
  Cancelled: 'gray',
  Reversed:  'rose',
};

export function getPaymentMethodLabel(method: string): string {
  return PAYMENT_METHOD_LABELS[method as PaymentMethod] ?? method ?? 'Unknown';
}

export function getPaymentStatusLabel(status: string): string {
  return status ?? 'Unknown';
}